File: script/maple/core/static_info.py
"""Copyright 2011 The University of Michigan

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

Authors - Jie Yu (jieyu@umich.edu)
"""

import os
from maple.core import logging
from maple.core import proto

def static_info_pb2():
    return proto.module('core.static_info_pb2')

class Image(object):
    def __init__(self, proto, db):
        self.proto = proto
        self.db = db
        self.offset_map = {}
    def id(self):
        return self.proto.id
    def name(self):
        return self.proto.name
    def basename(self):
        return os.path.basename(self.proto.name)
    def shortname(self):
        bn = self.basename()
        n1 = bn.split('.')[0]
        n2 = bn.split('-')[0]
        if len(n1) < len(n2):
            return n1
        else:
            return n2
    def add_inst(self, inst):
        self.offset_map[inst.offset()] = inst
    def __str__(self):
        content = []
        content.append('%-2d' % self.id())
        content.append('%s' % self.name())
        return ' '.join(content)

class Inst(object):
    def __init__(self, proto, db):
        self.proto = proto
        self.db = db
    def id(self):
        return self.proto.id
    def image(self):
        return self.db.image_map[self.proto.image_id]
    def offset(self):
        return self.proto.offset
    def debug_info(self):
        if not self.proto.HasField('debug_info'):
            return ''
        else:
            file = os.path.basename(self.proto.debug_info.file_name)
            line = self.proto.debug_info.line
            content = []
            content.append('%s' % file)
            content.append('+%d' % line)
            return ' '.join(content)
    def __str__(self):
        content = []
        content.append('%-5d' % self.id())
        content.append('%-10s' % self.image().shortname())
        content.append('0x%-6x' % self.offset())
        content.append('%s' % self.debug_info())
        return ' '.join(content)

class StaticInfo(object):
    def __init__(self):
        self.proto = static_info_pb2().StaticInfoProto()
        self.image_map = {}
        self.inst_map = {}
    def load(self, db_name):
        if not os.path.exists(db_name):
            return
        f = open(db_name, 'rb')
        self.proto.ParseFromString(f.read())
        f.close()
        for image_proto in self.proto.image:
            image = Image(image_proto, self)
            self.image_map[image.id()] = image
        for inst_proto in self.proto.inst:
            inst = Inst(inst_proto, self)
            self.inst_map[inst.id()] = inst
            self.image_map[inst.image().id()].add_inst(inst)
    def find_image(self, image_id):
        return self.image_map[image_id]
    def find_inst(self, inst_id):
        return self.inst_map[inst_id]
    def display_image_table(self, f):
        for image in self.image_map.itervalues():
            f.write(str(image))
            f.write('\n')
    def display_inst_table(self, f):
        for inst in self.inst_map.itervalues():
            f.write(str(inst))
            f.write('\n')

File: src/core/static_info.cc
// Copyright 2011 The University of Michigan
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Authors - Jie Yu (jieyu@umich.edu)

// File: core/static_info.cc - Static information for program binary or
// library binaries.

#include "core/static_info.h"

#include <fstream>
#include <sstream>

Inst *Image::Find(address_t offset) {
  InstAddrMap::iterator found = inst_offset_map_.find(offset);
  if (found == inst_offset_map_.end())
    return NULL;
  else
    return found->second;
}

bool Image::IsCommonLib() {
  if ((name().find("libc") != std::string::npos) ||
      (name().find("libpthread") != std::string::npos) ||
      (name().find("ld-") != std::string::npos) ||
      (name().find("libstdc++") != std::string::npos) ||
      (name().find("libgcc_s") != std::string::npos) ||
      (name().find("libm") != std::string::npos) ||
      (name().find("libnsl") != std::string::npos) ||
      (name().find("librt") != std::string::npos) ||
      (name().find("libdl") != std::string::npos) ||
      (name().find("libz") != std::string::npos) ||
      (name().find("libcrypt") != std::string::npos) ||
      (name().find("libdb") != std::string::npos) ||
      (name().find("libexpat") != std::string::npos) ||
      (name().find("libbz2") != std::string::npos))
    return true;
  else
    return false;
}

bool Image::IsLibc() {
  if (name().find("libc") != std::string::npos)
    return true;
  else
    return false;
}

bool Image::IsPthread() {
  if (name().find("libpthread") != std::string::npos)
    return true;
  else
    return false;
}

std::string Image::ShortName() {
  size_t found = name().find_last_of('/');
  if (found != std::string::npos)
    return name().substr(found + 1);
  else
    return name();
}

void Image::Register(Inst *inst) {
  inst_offset_map_[inst->offset()] = inst;
}

void Inst::SetDebugInfo(const std::string &file_name, int line, int column) {
  DebugInfoProto *di_proto = proto_->mutable_debug_info();
  di_proto->set_file_name(file_name);
  di_proto->set_line(line);
  di_proto->set_column(column);
}

std::string Inst::DebugInfoStr() {
  if (!HasDebugInfo()) {
    return "";
  } else {
    const DebugInfoProto &di_proto = proto_->debug_info();
    size_t found = di_proto.file_name().find_last_of('/');
    std::stringstream ss;
    if (found != std::string::npos)
      ss << di_proto.file_name().substr(found + 1);
    else
      ss << di_proto.file_name();
    ss << " +" << std::dec << di_proto.line();
    return ss.str();
  }
}

std::string Inst::ToString() {
  std::stringstream ss;
  ss << std::hex << id() << " " << image_->ToString() << " 0x" << offset();
  if (HasDebugInfo())
    ss << " (" << DebugInfoStr() << ")";
  return ss.str();
}

StaticInfo::StaticInfo(Mutex *lock)
    : lock_(lock),
      curr_image_id_(0),
      curr_inst_id_(0) {
  // empty
}

Image *StaticInfo::CreateImage(const std::string &name) {
  ImageProto *image_proto = proto_.add_image();
  image_id_type image_id = GetNextImageID();
  image_proto->set_id(image_id);
  image_proto->set_name(name);
  Image *image = new Image(image_proto);
  image_map_[image_id] = image;
  return image;
}

Inst *StaticInfo::CreateInst(Image *image, address_t offset) {
  InstProto *inst_proto = proto_.add_inst();
  inst_id_type inst_id = GetNextInstID();
  inst_proto->set_id(inst_id);
  inst_proto->set_image_id(image->id());
  inst_proto->set_offset(offset);
  Inst *inst = new Inst(image, inst_proto);
  inst_map_[inst_id] = inst;
  image->Register(inst);
  return inst;
}

Image *StaticInfo::FindImage(const std::string &name) {
  for (ImageMap::iterator it = image_map_.begin();
       it != image_map_.end(); ++it) {
    Image *image = it->second;
    if (image->name().compare(name) == 0)
      return image;
  }
  return NULL;
}

Image *StaticInfo::FindImage(image_id_type id) {
  ImageMap::iterator it = image_map_.find(id);
  if (it == image_map_.end())
    return NULL;
  else
    return it->second;
}

Inst *StaticInfo::FindInst(inst_id_type id) {
  InstMap::iterator it = inst_map_.find(id);
  if (it == inst_map_.end())
    return NULL;
  else
    return it->second;
}

void StaticInfo::Load(const std::string &db_name) {
  std::fstream in(db_name.c_str(), std::ios::in | std::ios::binary);
  proto_.ParseFromIstream(&in);
  in.close();
  // setup image map
  for (int i = 0; i < proto_.image_size(); i++) {
    ImageProto *image_proto = proto_.mutable_image(i);
    Image *image = new Image(image_proto);
    image_id_type image_id = image->id();
    image_map_[image_id] = image;
    if (image_id > curr_image_id_)
      curr_image_id_ = image_id;
  }
  // setup inst map
  for (int i = 0; i < proto_.inst_size(); i++) {
    InstProto *inst_proto = proto_.mutable_inst(i);
    Image *image = FindImage(inst_proto->image_id());
    Inst *inst = new Inst(image, inst_proto);
    inst_id_type inst_id = inst->id();
    inst_map_[inst_id] = inst;
    image->Register(inst);
    if (inst_id > curr_inst_id_)
      curr_inst_id_ = inst_id;
  }
}

void StaticInfo::Save(const std::string &db_name) {
  std::fstream out(db_name.c_str(),
                   std::ios::out | std::ios::trunc | std::ios::binary);
  proto_.SerializeToOstream(&out);
  out.close();
}

